pkg/storegateway/bucket.go

```diff
@@ -731,17 +963,26 @@ func (s *BucketStore) streamingSeriesSetForBlocks(
 	chunksLimiter ChunksLimiter, // Rate limiter for loading chunks.
 	seriesLimiter SeriesLimiter, // Rate limiter for loading series.
 	stats *safeQueryStats,
-) (storepb.SeriesSet, *hintspb.SeriesResponseHints, error) {
+	reusePostings [][]storage.SeriesRef,      // Used if not empty.
+	reusePendingMatchers [][]*labels.Matcher, // Used if not empty.
+) (storepb.SeriesSet, seriesChunksSetIterator, *hintspb.SeriesResponseHints, error) {
```

Comment:
This function currently serves 3 distinct purposes. Do they need to be the same function or can we split them apart?

* streaming series
	* doesn't take postings, pending matchers, chunk readers or chunk limiter
	* returns postings, a SeriesSet
* streaming chunks
	* takes postings, pending matchers, chunk readers or chunk limiter 
	* returns seriesChunksSetIterator
* non-streaming chunks&series
	* the same version as it is on main now

All of this will also make it easier to remove either streaming or non-streaming versions later on with less code restructure.

I think it also makes sense to move hints calculation out of the three (it's only iterating over blocks and adding each to the SeriesResponseHints).